server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java

```diff
@@ -0,0 +1,259 @@
+        this.settings = settings;
+        this.clusterService = clusterService;
+        this.client = client;
+        this.healthService = healthService;
+        this.clock = Clock.systemUTC();
+        this.pollInterval = HEALTH_PERIODIC_LOGGER_POLL_INTERVAL_SETTING.get(settings);
+    }
+
+    /**
+     * Initializer method to avoid the publication of a self reference in the constructor.
+     */
+    public void init() {
+        clusterService.addListener(this);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(HEALTH_PERIODIC_LOGGER_POLL_INTERVAL_SETTING, this::updatePollInterval);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        DiscoveryNode healthNode = HealthNode.findHealthNode(event.state());
+        if (healthNode == null) {
+            this.isHealthNode = false;
+            this.cancelJob();
+            return;
+        }
+        final boolean isCurrentlyHealthNode = healthNode.getId().equals(this.clusterService.localNode().getId());
+        if (this.isHealthNode != isCurrentlyHealthNode) {
+            this.isHealthNode = isCurrentlyHealthNode;
+            if (this.isHealthNode) {
+                // we weren't the health node, and now we are
+                maybeScheduleJob();
+            } else {
+                // we were the health node, and now we aren't
+                cancelJob();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        SchedulerEngine engine = scheduler.get();
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(HEALTH_PERIODIC_LOGGER_JOB_NAME)) {
+            if (this.isHealthNode && this.currentlyRunning.compareAndExchange(false, true) == false) {
+                this.callGetHealth();
+            }
+        }
+    }
+
+    // default visibility for testing purposes
+    void callGetHealth() {
+        this.healthService.getHealth(this.client, null, false, 0, this.resultsListener);
+    }
+
+    // default visibility for testing purposes
+    SchedulerEngine getScheduler() {
+        return this.scheduler.get();
+    }
+
+    /**
+     * Handle the result of the Health Service getHealth call
+     */
+    // default visibility for testing purposes
+    final ActionListener<List<HealthIndicatorResult>> resultsListener = new ActionListener<List<HealthIndicatorResult>>() {
+        @Override
+        public void onResponse(List<HealthIndicatorResult> healthIndicatorResults) {
+            HealthPeriodicLoggerResult result = new HealthPeriodicLoggerResult(healthIndicatorResults);
+            Map<String, Object> resultsMap = result.toMap();
```

Comment:
It feels that the `HealthPeriodicLoggerResult` is only used to convert the list of indicator results to a map. Is it worth creating an object just for this?

Should we inline the method or make it static and remove the `HealthPeriodicLoggerResult`? 

What do you think?

Reply:
Good observation. I've turned it into a method on the `HealthPeriodicLogger` class.